tsdb/store.go

```diff
@@ -425,18 +448,28 @@ func (s *Store) DeleteDatabase(name string) error {
 // provided retention policy, remove the retention policy directories on
 // both the DB and WAL, and remove all shard files from disk.
 func (s *Store) DeleteRetentionPolicy(database, name string) error {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
 	// Close and delete all shards under the retention policy on the
 	// database.
-	for shardID, sh := range s.shards {
+	responses := s.executeInParallel(func(shardID uint64, sh *Shard) func() error {
 		if sh.database == database && sh.retentionPolicy == name {
-			// Delete the shard from disk.
-			if err := s.deleteShard(shardID); err != nil {
-				return err
+			return func() error {
+				// Delete the shard from disk.
+				return s.DeleteShard(shardID)
```

Comment:
This is needed to avoid a data race, but I'm not sure how useful this is other than just ensuring the database doesn't get locked up while a retention policy is being removed from shards. This takes a lock on the `tsdb.Store` instance so there's no way that this can be run in parallel.